cluster/v1alpha1/types_rolloutstrategy.go

```diff
@@ -0,0 +1,103 @@
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// +k8s:deepcopy-gen=true
+
+// RolloutStrategy API used by workload applier APIs to define how the workload will be applied to the selected clusters by the Placement and DecisionStrategy.
+
+const (
+	//All means apply the workload to all clusters in the decision groups at once.
+	All string = "All"
+	//Progressive means apply the workload to the selected clusters progressively per cluster.
+	Progressive string = "Progressive"
+	//ProgressivePerGroup means apply the workload to the selected clusters progressively per group.
+	ProgressivePerGroup string = "ProgressivePerGroup"
+)
+
+// Rollout strategy to apply workload to the selected clusters by Placement and DecisionStrategy.
+type RolloutStrategy struct {
+	// Rollout strategy Types are All, Progressive and ProgressivePerGroup
+	// 1) All means apply the workload to all clusters in the decision groups at once.
+	// 2) Progressive means apply the workload to the selected clusters progressively per cluster. The workload will not be applied to the next cluster unless one of the current applied clusters reach the successful state or timeout.
+	// 3) ProgressivePerGroup means apply the workload to decisionGroup clusters progressively per group. The workload will not be applied to the next decisionGroup unless all clusters in the current group reach the successful state or timeout.
+	// +kubebuilder:validation:Enum=All;Progressive;ProgressivePerGroup
+	// +kubebuilder:default:=All
+	// +optional
+	Type string `json:"type,omitempty"`
+
+	// All define required fields for RolloutStrategy type All
+	// +optional
+	All *RolloutAll `json:"all,omitempty"`
+
+	// Progressive define required fields for RolloutStrategy type Progressive
+	// +optional
+	Progressive *RolloutProgressive `json:"progressive,omitempty"`
+
+	// ProgressivePerGroup define required fields for RolloutStrategy type ProgressivePerGroup
+	// +optional
+	ProgressivePerGroup *RolloutProgressivePerGroup `json:"progressivePerGroup,omitempty"`
```

Comment:
Is there a reason behind having these as separate keys? Could it not be streamlined as providing the `Type` and inlining all of the options? If that's an option, I can work on a PR. At minimum, I think it might be nice to have `Timeout` at the same level as `Type`?

```golang
type RolloutStrategy struct {
	// +kubebuilder:validation:Enum=All;Progressive;ProgressivePerGroup
	// +kubebuilder:default:=All
	// +optional
	Type string `json:"type,omitempty"`

	// +optional
	Timeout `json:",inline"`

	// +optional
	MandatoryDecisionGroups `json:",inline"`

	// +kubebuilder:validation:Pattern="^((100|[0-9]{1,2})%|[0-9]+)$"
	// +kubebuilder:validation:XIntOrString
	// +optional
	MaxConcurrency intstr.IntOrString `json:"maxConcurrency,omitempty"`
}
```

Reply:
this is designed as an union type, a certain type is point to the a certain field. see here https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md. This could avoid unexpected wrong configurations. For instance, you would not want user set Type to Progressive and still set the MandatoryDecisionGroups.

Reply:
It has been done with union type to avoid miss leading the endUser as for ex; when the Type is ALL there is no need to define the MandatoryDecisionGroups and MaxConcurrency. Another example; when the Type is ProgressivePerGroup no need to define MaxConcurrency. 
So make every Type has its own field make more sense.

Reply:
Actually, I don't see the union type even implemented yet in Kubernetes? Kubebuilder doesn't support `OneOf` (yet?), but some Kustomize post-processing on the CRD could let consumers use the `OneOf` in the CRD rather than having a `Type` field. This would also avoid confusion since it'd be impossible to set the `Type` to one thing and then provide a set of `Progressive` options (as I think can be done as it is now)?

```yaml
oneOf:
- required: [ "all" ]
- required: [ "progressive" ]
- required: [ "progressivePerGroup" ]
```

Though without Kubebuilder support, it'd have to be implemented by the consumers of the library, unfortunately.